### ISSUE TITLE
gitbase integration

### ISSUE THREAD

**Issue by ajnavarro** [src-d/jgit-spark-connector#345]
To have an idea of how gitbase will work with Spark, we need to start some integration with both projects

- Open a new branch to start working on it
- Change the schema to the gitbase one
- Joins should be pushed down as usual, but also the special UDFs (history_idx, commit_has_blob, and so on)
- Remove as much code as possible
- Modify tests as less as possible

With this integration, all the iterators should be removed, and the way to access data will be making an sql query from the pushed down data, and send it using mysql driver to gitbase.

**Comment by ajnavarro** [comment]
To keep it simple and to have something working asap due the gitbase limitations, as first step we can just use the JDBC spark connector to connect spark with gitbase.

**Comment by erizocosmico** [comment]
Created the `gitbase` branch on engine to merge pull requests from this feature there.

**Comment by erizocosmico** [comment]
We've encountered a few problems with this task that we did not foresee during the planning of it.

### The problem

Because of gitbase's data model we need to make use of the `history_idx`,  `commit_has_blob` and `commit_has_tree` functions in order to be able to perform join operations between several tables, which is crucial to make the engine work with this.
Since the JDBC datasource does not do pushdown of these functions they must be implemented in Spark, and we cannot do this, since Spark does not have the repositories to compute these relationships.
In summary, we cannot use the JDBC datasource to do this task. At least as long as the data model inside gitbase remains the same.

### A possible solution

We've come with a solution to make this work, but it would need to be done on gitbase's side: implementing views. These views would be the tables but already joined, so the JDBC datasource can just fullscan the tables and operate normally.

### TL;DR

So we either need to change the data model of gitbase to not make use of these functions or we need to have views implemented with these pre-joined tables as views if we want to do this simple integration with the JDBC driver.

### The complex approach

Another way more complex option is to do a _not so simple_ integration, which would consist in implementing `CatalystScan`, the UDFs, some rules to merge tables into just one, pushing down the joins and then with all that build a query that would be sent to gitbase to retrieve the data with just the plain SQL.

**Comment by smola** [comment]
https://github.com/src-d/gitbase-spark-connector
